backend/main/chapters/c09_combining_booleans.py

```diff
@@ -0,0 +1,638 @@
+Thankfully Python allows us to write a single, long expression on multiple lines,
+as long as it is within parentheses.
+We can go to the next line right after a boolean operator.
+Type and run this code in the editor:
+
+    print(True or
+    False and
+    False)
+        """
+        program_in_text = False
+        program = "print(True or False and False)"
+
+    final_text = """
+Remember to always wrap your multi-line expressions inside parentheses to be safe.
+    """
+
+
+class CombiningAndWithOr(Page):
+    class CheckTheWholeBoard(VerbatimStep):
+        """
+The tic-tac-toe board can be represented by 3 rows.
+If we have a winning row, it's either row 1, OR row 2, OR row 3.
+For each row, we can reuse our row-checking code from before
+(with parentheses around it):
+
+    (row[0] == row[1] and row[0] == row[2])
+
+To fit all 3 row checks on multiple lines,
+we need one extra set of parentheses around the whole thing, after `if`.
+We break lines after the boolean operators `or` which are connecting the bigger blocks.
+Type and run this code, make sure you understand the complex boolean expression:
+
+__program_indented__
+        """
+        def program(self):
+            def is_winning_board(row1, row2, row3):
+                if (
+                    (row1[0] == row1[1] and row1[0] == row1[2]) or
+                    (row2[0] == row2[1] and row2[0] == row2[2]) or
+                    (row3[0] == row3[1] and row3[0] == row3[2])
+                ):
+                    print("We have a winner!")
+
+            row1 = ["X", "O", "X"]
+            row2 = ["X", "X", "X"]
+            row3 = ["O", "O", "X"]
+
+            is_winning_board(row1, row2, row3)
+
+    class DiagonalWinnerExercise(ExerciseStep):
+        """
+Exercise: write a function `diagonal_winner` to check if there's a diagonal winner on a tic-tac-toe board.
+As before, the board is given as 3 rows: `row1, row2, row3`, which the function takes as arguments.
+Print out `X wins!` or `O wins!` depending on which player has a winning diagonal.
+Otherwise print `No winners.`
+For example, if
+
+    row1 = ["X", "O", "X"]
+    row2 = ["X", "X", "O"]
+    row3 = ["O", "O", "X"]
+
+then `diagonal_winner(row1, row2, row3)` should print `X wins!`
+        """
+        hints = """
+How many diagonals are there on the board? What do they have in common?
+Which entries of `row1, row2, row3` make up each diagonal?
+Similar to how we checked a row, check that the 3 entries on a diagonal are equal to each other, using `and`.
+Similar to how we checked three rows together, check the two diagonals together, using `or`.
+As before, use a multi-line boolean expression where `if` contains lines that are connected with `or`. Only 2 lines 
+needed.
+Don't forget the `else` part!
+        """
+
+        def solution():
+            @returns_stdout
+            def diagonal_winner(_, row1: List[str], row2: List[str], row3: List[str]):
```

Comment:
Type annotations aren't enough here, you will need to override `generate_inputs`.

Reply:
Would this work:
```python
        def generate_inputs(cls):
            return {
                "row1": [random.choice(["X", "O"]) for _ in range(3)],
                "row2": [random.choice(["X", "O"]) for _ in range(3)],
                "row3": [random.choice(["X", "O"]) for _ in range(3)]
            }
```

Reply:
I think so, and yes to `@classmethod`.

To see this in action, try cheating by submitting a solution that passes the manual `tests` just using hardcoded if statements. See the difference with and without `generate_inputs`.

Reply:
Oh I see. Random testing is necessary. I'll make the changes and commit this.